radix-engine/src/wasm/wasmi.rs

```diff
@@ -11,279 +14,478 @@ use crate::wasm::constants::*;
 use crate::wasm::errors::*;
 use crate::wasm::traits::*;
 
+type HostState = WasmiInstanceEnv;
+
+/// A `WasmiModule` defines a parsed WASM module "template" Instance (with imports already defined) and Store,
+/// which keeps user data.
+/// "Template" (Store, Instance) tuple are cached together, and never to be invoked.
+/// Upon instantiation Instance and Store are cloned, so the state is not shared between instances.
+/// It is correctly `Send + Sync` - which is good, because this is the thing which is cached in the
+/// ScryptoInterpreter caches.
 pub struct WasmiModule {
-    module: Module,
+    store: Store<HostState>,
+    instance: Instance,
```

Comment:
Could we call this eg `template_store` and `template_instance` to indicate that this is the template which should be cloned? 👍